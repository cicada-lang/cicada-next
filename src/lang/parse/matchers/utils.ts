import * as Exps from "../../exp"

export function piBindingtoFnBindingFrom(binding: Exps.PiBinding): Exps.FnBinding {
  switch (binding.kind) {
    case "PiBindingNameless": {
      return Exps.FnBindingName("_")
    }

    case "PiBindingNamed": {
      return Exps.FnBindingName(binding.name)
    }

    case "PiBindingImplicit": {
      return Exps.FnBindingImplicit(binding.name)
    }
  }
}
