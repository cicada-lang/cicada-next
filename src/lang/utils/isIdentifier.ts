const identifierRegExp = new RegExp("^([-_\\p{Letter}][-_\\p{Letter}0-9]*)$", "u")

export function isIdentifier(text: string): boolean {
  return identifierRegExp.test(text)
}
