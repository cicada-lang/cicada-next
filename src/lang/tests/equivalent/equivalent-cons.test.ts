import { expect, test } from "vitest"
import { expectCodeToFail, runCode } from "../utils"

test("equivalent Cons", async () => {
  const output = await runCode(`

function id(T: Type, x: T): T {
  return x
}

compute equivalent exists (T: Type) T {
    cons(Trivial, sole)
  = cons(Trivial, id(Trivial, sole))
}

`)

  expect(output).toMatchSnapshot()
})

test("equivalent Cons -- fail", async () => {
  await expectCodeToFail(`

compute equivalent exists (T: Type) T {
    cons(String, "abc")
  = cons(String, "xyz")
}

`)
})
