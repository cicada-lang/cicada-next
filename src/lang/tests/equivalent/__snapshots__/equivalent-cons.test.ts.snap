// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`equivalent Cons 1`] = `"refl(implicit exists (Type) T1, implicit cons(Trivial, sole)): Equal(exists (Type) T1, cons(Trivial, sole), cons(Trivial, sole))"`;
